Example of benchmarking scan backends against pure implementations in the respective backend

Dekermanjian opened the pull request:
added trust_inputs to pytensor functions, added multiple pytensor backends, added block_until_ready to jax calls

<!-- !! Thank your for opening a PR !! -->

<!--- Provide a self-contained summary of your changes in the Title above -->
<!--- This is what will be shown in the automatic release notes: https://github.com/pymc-devs/pytensor/releases -->

## Description
This PR adds a benchmarking notebook that benchmarks several algorithms in different backends and compares them to pure implementations in their respective backend.

## Related Issue
<!--- It is good practice to first open an issue explaining the bug / new feature that is addressed by this PR -->
<!--- Please type an `x` in one of the boxes below and provide the issue number after the # sign: -->
- [ x] Related to #1631 

## Checklist
<!--- Make sure you have completed the following steps before submitting your PR -->
<!--- Feel free to type an `x` in all the boxes below to let us know you have completed the steps: -->
- [x ] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)
<!--- You may find this guide helpful: https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/ -->

## Type of change
<!--- Select one of the categories below by typing an `x` in the box -->
- [ ] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ x] Other (please specify): Example Notebook
<!--- Additionally, if you are a maintainer or reviewer, please make sure that the appropriate labels are added to this PR -->


<!-- readthedocs-preview pytensor start -->
----
📚 Documentation preview 📚: https://pytensor--1632.org.readthedocs.build/en/1632/

<!-- readthedocs-preview pytensor end -->

Dekermanjian committed:
added trus_inputs to pytensor functions, added multiple pytensor backends, added block_until_ready to jax calls

Dekermanjian committed:
updated renderer for reviewnb

Dekermanjian committed:
Updated how jax functions are blocked and updated jax fibonacci algorithm to prevent constant folding

ricardoV94 committed:
Add benchmark notebook focusing on numba scan

ricardoV94 commented:
@Dekermanjian I pushed a notebook building on top of yours. I focused on a different variation of scan, where the n_steps are constant, but b is symbolic. This is because right now we have a very complex shape graph for symbolic steps that just makes things messier, and also prevents proper inplace sometimes.
Related issues: #1283,  #112
Related PRs: #1299

On my machine, the function evaluation costs:
c-backend: 46.us
numba-backend: 5.5 us
numba-backend-skipping-pytensor: 3.3 us
direct_numba: 1.2 us
numba_that_looks_like_pytensor: 2.2 us

My hypothesis for why `numba-backend-skipping-pytensor` is slower than `numba_that_looks_like_pytensor` is because of how we implement Elemwise, which I didn't try to replicate.

The difference between `numba-backend` and `numba-backend-skipping-pytensor` should come further down after #1351

ricardoV94 commented:
Small discrepancy that I don't know if it matters in your original notebook. You are using `pt.constant(1.0)` which is float by default, but you use int32 in the numba/jax variations.

ricardoV94 commented:
> I replaced the for loop with a fori_loop to prevent the constant folding.

What makes you think fori wouldn't constant_fold but scan would? The point @aseyboldt was trying to make is that for jax the number of steps is always constant (even if you pass it at runtime, it's going to specialize on the specific value), and if you have constant a/b then all the variables are constant as far as JAX can see. The trick to use symbolic a/b is that those aren't treated as constant by jax and therefore it won't constant-fold it. You can always inspect the jaxepr or whatever it's called to be sure though.

Anyway I prefer the constant n_steps because of the known issues related to Scan that I linked to above.

ricardoV94 committed:
Add global N_STEPS

ricardoV94 commented:
> I'm a bit baffled that pure numba remains so fast for n=1000, I guess it just manages to stay in scalar land

Yeah, that explains a big factor: 
<img width="679" height="695" alt="image" src="https://github.com/user-attachments/assets/0083e3b7-9e78-4fd3-8052-0365f9fbf603" />

ricardoV94 committed:
Try a scalar/array variants of the pure numba impl

ricardoV94 commented:
Your second example is not actually testing scan. PyTensor as a rewrite that converts pure elementwise scans to regular elementwise. You should always `dprint` your final version to be sure you're testing what you mean to.

<img width="561" height="790" alt="image" src="https://github.com/user-attachments/assets/a07b9988-dcda-47fd-8f0c-25264f5264f4" />

In this case you can keep it by excluding the specific rewrite: 
<img width="1063" height="662" alt="image" src="https://github.com/user-attachments/assets/06f5e889-08de-43c0-bccc-8c4422d86056" />

ricardoV94 commented:
Summary from the fibonacci bit on numba. We're not doing anything dumb, the difference is we were basically comparing a scalar loop (due to numba bug) and an array loop. This performance question can be more properly addressed by rewriting scalar scan loops as scalar loops, as mentioned in #301.

More broadly, we could try to modify the scan inner function to accept scalars as inputs and return scalars as outputs. Even more broadly we could try to create a numba function that will store the results in the output buffer like scan tries in the C backend. This is tricky because we have implemented Numba Ops in a way that they don't receive output storage, and from what I see here numba / llvm isn't clever enough to avoid useless intermediate copies:

```python
import numpy as np
import numba

@numba.njit
def foo(x):
    for i in range(x.shape[0]):
        y = x[i:i+1]
        y *= -1
        x[i:i+1] = y
    return x

@numba.njit
def bar(x):
    for i in range(x.shape[0]):
        y = x[i:i+1]
        y *= -1    
    return x

x = np.ones(50)
assert foo(x.copy()).sum() == bar(x.copy()).sum() == -50.0
%timeit foo(x)  # 2.1 us
%timeit bar(x)  # 600 ns
```

Note that we do have inplace rewrites in the scan dispatch, so output buffers aren't always created. However this doesn't prevent the copy back to the original array.

Dekermanjian committed:
updated pytensor elemwise to exclude rewrite that removes scan, updated fori_loop to explicitly tell it not to constant fold

Dekermanjian committed:
updated jax fibonacci to ensure no constant folding. ensured all pytensor dprint included scan in the graph

Dekermanjian committed:
updated pytensor scan methods to always include fixed iterations in n_step arg, updated jax elemntwise to use fori_loop

Dekermanjian commented:
@ricardoV94 I updated the Pytensor fibonacci algorithm to follow your method of fixing the number of iterations in the scans `n_step` argument. I wasn't sure if this was also needed when `sequences` were passed to `pytensor.scan` but I also fixed the number of iterations in those cases as well, just in case.

ricardoV94 commented:
> @ricardoV94 I updated the Pytensor fibonacci algorithm to follow your method of fixing the number of iterations in the scans `n_step` argument. I wasn't sure if this was also needed when `sequences` were passed to `pytensor.scan` but I also fixed the number of iterations in those cases as well, just in case.

When sequences are passed n_steps defaults to those, but it doesn't hurt to specify again (or shouldn't!)

Dekermanjian committed:
moved alarm checks outside of loop for numba CUSUM

ricardoV94 commented:
I want to rerun some, as some of the differences are rather wild. 

One thing is the Fibonacci in jax should have fori_loop. All jax examples should use scan or fori, otherwise it's compiling an insane unrolled loop graph that nobody would write in practice.

Also 100k is a bit overkill, why did you go so large?

We should looks at some of these in a hackathon to try some ideas on how to bridge the gap. We know the Fibonacci is a difference between scalar and array operations, numba was kind of cheating there.

I also opened some recent issues on scan related optimizations. We should open more as we find actionable strategies.

One I didn't open yet is allowing giving an `out` parameter for inner graphs to store their outputs in, to avoid an extra level of copy we are doing compared to your direct implementations

Dekermanjian commented:
> I want to rerun some, as some of the differences are rather wild.

Yes, I totally understand. I could've easily messed something up with the implementation of the algorithms!

> One thing is the Fibonacci in jax should have fori_loop. All jax examples should use scan or fori, otherwise it's compiling an insane unrolled loop graph that nobody would write in practice.

 Okay, I will make that change ASAP. 

> Also 100k is a bit overkill, why did you go so large?

No particular reason, I just wanted to make sure it was large enough to counter any sort of external overhead. I can bring it down to 10k or 1k. 

> We should looks at some of these in a hackathon to try some ideas on how to bridge the gap. We know the Fibonacci is a difference between scalar and array operations, numba was kind of cheating there.

That sounds like a good idea. When you say hackathon is that just getting on a call and walking through it? Do you think it is worth creating a different fibonacci function where one runs on scalar operations and the other runs on array operations?

> I also opened some recent issues on scan related optimizations. We should open more as we find actionable strategies.

> One I didn't open yet is allowing giving an `out` parameter for inner graphs to store their outputs in, to avoid an extra level of copy we are doing compared to your direct implementations

 How big of an impact are the additional levels of copy? Does it scale with the number of iterations/size of the array? Does it also impact memory usage?

Dekermanjian committed:
updated fibonacci jax to use fori_loop and reduced N_STEPS
